Source/WebCore/rendering/RenderLayer.cpp

```diff
@@ -842,12 +866,16 @@ void RenderLayer::collectLayers(std::unique_ptr<Vector<RenderLayer*>>& positiveZ
     // Overflow layers are just painted by their enclosing layers, so they don't get put in zorder lists.
     bool includeHiddenLayer = (m_hasVisibleContent || m_intrinsicallyComposited) || ((m_hasVisibleDescendant || m_hasIntrinsicallyCompositedDescendants) && isStacking);
     includeHiddenLayer |= page().hasEverSetVisibilityAdjustment();
-    if (includeHiddenLayer && !isNormalFlowOnly()) {
-        auto& layerList = (zIndex() >= 0) ? positiveZOrderList : negativeZOrderList;
-        if (!layerList)
-            layerList = makeUnique<Vector<RenderLayer*>>();
-        layerList->append(this);
-        accumulatedDirtyFlags.add(m_compositingDirtyBits);
+    if (!isNormalFlowOnly()) {
+        if (includeHiddenLayer) {
+            auto& layerList = (zIndex() >= 0) ? positiveZOrderList : negativeZOrderList;
+            if (!layerList)
+                layerList = makeUnique<Vector<RenderLayer*>>();
+            layerList->append(this);
+            accumulatedDirtyFlags.add(m_compositingDirtyBits);
+            setWasIncludedInZOrderTree();
```

Comment:
It's a bit odd to set/clear these bits here. I'm not convinced that they are always updated correctly in the face of z-order tree mutations. I'd expect them to be dirtied via something closer to `m_visibleContentStatusDirty`.

Reply:
I'm less convinced of that. The logic for whether we include/omit layers is here, and takes into account more than just visibility state (the intrinsic compositing checks), so we'd need lots of dirtying all over the place otherwise.

Reply:
I guess the relevant dirty bit here is the existing z-order lists dirty state. That does already get set when we make something visible, though I think it's possibly not set in some reparenting cases.

That's an existing bug that would affect shipping, but I can try make a test case and fix it too.

Reply:
Ok, got a test that had a failure, and updated to fix that too.